comments/heroku_logs_to_cloudwatch.md

```diff
@@ -0,0 +1,79 @@
+
+Given that we already use AWS for a number of functions, if we want long-term log storage it makes sense to write a Lambda function that will ingest the logs into AWS Cloudwatch, which is Amazon's logging and metrics service.
+
+### Why does this document exist?
+
+Because while there are a few pieces of code in public Github repositories that provide a basic example of how shipping logs from Heroku to Cloudwatch via Lambda works (see [here](https://github.com/rwilcox/heroku_cloudwatch_sync/blob/master/src/heroku_sync_to_cloudwatch.py), there are others available with some basic googling), attempting to adapt this example code into an actual working Lambda function that didn't scream blue murder in our alerting channel took four rounds of refactoring. The learnings are too numerous to be captured in comments in the function code, so I'm writing them down here.
+
+## Basic operation
+
+In theory, very simple:
+
+- Write a Lambda function handler in Python capable of parsing Heroku log data and forwarding it on to Cloudwatch
+- Connect Heroku up to the Lambda function by using the Heroku CLI to add the Lambda function URL as a log drain: `heroku drains:add <function url> -a <appname>`.
+
+In practice, very difficult.
+
+### Problem 1: Cloudwatch log groups and log streams
+
+AWS Cloudwatch logging relies on the concept of log *streams*, which contain streams of individual log messages, which are all part of a single log *group*. Log streams are the things that actually get written to by the Lambda function, and the log group allows filtering and searching to be run across multiple log streams at once.
+
+Most of the example code for pushing logs to Cloudwatch only pushed the logs to a single static log stream. This unfortunately does not work in practice, because of...
+
+### Problem 2: Parallel Lambda function execution
+
+If you push five separate events up to a Lambda function, AWS will not invoke one instance of the function that processes those five events one after another. Instead it will spawn five function instances that each process one event in parallel. 
+
+This means that we potentially have five function instances all simultaneously attempting to write to the same Cloudwatch log stream. This was the cause of most of the initial errors to do with our implementation, thanks to...
+
+### Problem 3: Cloudwatch log stream sequence tokens
+
+When you do a `put_log_event` operation against a Cloudwatch log stream, one of the things you're supposed to pass (according to [boto3 documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/logs.html#CloudWatchLogs.Client.put_log_events)) is a sequence token. There is only one valid sequence token at any one time per log stream. Attempting to write to a log stream with an incorrect or out-of-date sequence token results in `InvalidSequenceTokenException`. And every *successful* write to a log stream updates the sequence token to a new value.
```

Comment:
The current AWS docs for `PutLogEvents` say otherwise:

https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html
> The sequence token is now ignored in `PutLogEvents` actions. `PutLogEvents` actions are always accepted and never return `InvalidSequenceTokenException` or `DataAlreadyAcceptedException` even if the sequence token is not valid. You can use parallel `PutLogEvents` actions on the same log stream.